expression/constant_propagation.go

```diff
@@ -138,10 +85,51 @@ func (s *propagateConstantSolver) propagateEQ() {
 	}
 }
 
+// propagateOthers propagates all deterministic conditions.
+// e.g. For expression a = b and b = c and c = d and c < 1 , we can get extra a < 1 and b < 1 and d < 1.
+// However, for a = b and a < rand(), we cannot propagate a < rand() to b < rand() because rand() is non-deterministic
+// We maintain a unionSet representing the equivalent for every two columns.
+func (s *propagateConstantSolver) propagateOthers() {
+	s.unionSet = &multiEqualSet{}
+	s.unionSet.init(len(s.columns))
+	for i := range s.conditions {
+		if fun, ok := s.conditions[i].(*ScalarFunction); ok && fun.FuncName.L == ast.EQ {
+			lCol, lOk := fun.GetArgs()[0].(*Column)
+			rCol, rOk := fun.GetArgs()[1].(*Column)
+			if lOk && rOk {
+				lID := s.getColID(lCol)
+				rID := s.getColID(rCol)
+				s.unionSet.addRelation(lID, rID)
+			}
+		}
+	}
+
+	condsLen := len(s.conditions)
+	for j, colj := range s.columns {
+		for k := j + 1; k < len(s.columns); k++ {
+			if s.unionSet.findRoot(j) != s.unionSet.findRoot(k) {
+				continue
+			}
+			colk := s.columns[k]
+			for i := 0; i < condsLen; i++ {
```

Comment:
s/ condsLen/ len(s.conditions) 
and remove line 107
since we only use it here.

Reply:
I think it's better to keep `condsLen` because we're adding elements to `conditions` while iterating it. Iteration on new elements are not needed.

Reply:
We need to comment for the behavior of the 3-layer loop.

Reply:
addressed, thanks